### DISCUSSION TITLE
Issue with GetSourceAsync()

### DISCUSSION THREAD

**Discussion by SilentStorm-1986** [cefsharp/CefSharp#4279] [General]
I am using WinForms Cef (latest build). I'm connecting via SSL proxy, which seems to load the page fine,  however I can't retrieve the page source with GetSourceAsync() or ViewSource() (empty string and empty txt in notepad respectively).

Here's my code:
```
Cef.UIThreadTaskFactory.StartNew(Sub()

                                                     Dim newrc = chromBrowser.GetBrowser.GetHost.RequestContext
                                                     Dim dict As New Dictionary(Of String, Object)
                                                     dict("mode") = "fixed_servers"
                                                     //SSL proxy, page loads fine
                                                     dict("server") = proxy & ":" & port
                                                     newrc.SetPreference("proxy", dict, [error])

                                                 End Sub)

chromBrowser.Load(strURL)

//set by EventHandler OnFrameLoadEnd
Do Until boolLoaded = True
    Application.DoEvents()
Loop

query = chromBrowser.GetSourceAsync()
query.Wait()
If query.Result IsNot Nothing Then
   //strHTML = ""
   strHTML = query.Result.ToString
End If
```

**Comment by SilentStorm-1986** [comment]
In an older build of Cef I have a similar issue. I get only the outer tags: `<html><head></head><body></body></html>`

I have no issues with any build when I'm using my IP and not a proxy.

**Reply by amaitland** [reply]
Assuming your proxy works correctly you need to wait for the proxy to be set use SetProxyAsync extension method and await it's result before calling Load. Use LoadUrlAsync instead of Load and await it's result before calling GetSourceAsync would be my recommendation. 

- http://cefsharp.github.io/api/105.3.x/html/Overload_CefSharp_RequestContextExtensions_SetProxyAsync.htm
- https://github.com/cefsharp/CefSharp/blob/a819fc4e4dbf2bede96a524c9fc68e8400a45591/CefSharp.Test/OffScreen/OffScreenBrowserBasicFacts.cs#L597
- https://github.com/cefsharp/CefSharp/blob/f84f3986cf20288940fb0589ac36c9272f75f46f/CefSharp.Test/Wpf/WpfBrowserBasicFacts.cs#L49

If you still have problems Check the log file for errors see https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#log-file

**Comment by SilentStorm-1986** [comment]
I fixed this the other day by using the following code, before calling GetSourceAsync():
```
Do Until chromBrowser.CanExecuteJavascriptInMainFrame = True
    Application.DoEvents()
Loop
```

Now I have a new problem with proxies that **will** connect through the cefclient.exe test build command line proxy switch, but NOT the RequestContext.SetPreference method, Cef.GetGlobalRequestContext.SetProxy method, RequestContext.Configure().WithProxyServer(proxy, port).Create() method OR settings.CefCommandLineArgs.Add("proxy-server", proxy) method.

I thought the error might have to do with the fixed_servers mode, however using a pac_url doesn't seem to work either.

Honestly I don't need to use these particular proxies, but I am curious as to why they work with the cefclient test build but won't work with the WinForms Cef assembly.

**Reply by amaitland** [reply]
Likely you've made a mistake somewhere. If you set via command line args you cannot change at runtime. 

You can use the xunit test at https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Test/OffScreen/OffScreenBrowserBasicFacts.cs#L562 modify the proxy and see if it runs.

**Comment by SilentStorm-1986** [comment]
I'm quite sure that I've set up the code correctly. All the aforementioned methods were tested independently from each other. I can switch the proxy from one list (the list that I got working the other day) to the other and the first will work and the other will not. 

Does cefclient dynamically run through different methods of connecting to the proxy until it has a connection? I set debug log to Verbose and cefclient, WinForms working proxy and WinForms failed proxy all show the exact same output. The proxy seems to connect but the page doesn't load.

The cefclient test build I'm using matches the latest WinForm version.

**Reply by amaitland** [reply]
Its my understanding proxy functionality is implemented entirely at the Chromium layer. All CEF based applications should behave the same.

**Comment by SilentStorm-1986** [comment]
I have one more piece of information. If I type the proxy address in the address bar in a browser, there seems to be a redirect prior to connection with the one that fails. Here's the server response for the proxy that fails:
`This is a proxy server, does not response to non-proxy requests`

Here's a typical response from a proxy that works:
```
Page failed to load.
URL: https://proxy
Error: ERR_SSL_PROTOCOL_ERROR (-107)
ERR_SSL_PROTOCOL_ERROR
```

**Comment by SilentStorm-1986** [comment]
I just solved it. I had to implement a RequestHandler class and leave the username & password as empty strings. Very strange but it worked.
